### PR TITLE
checks: stop getLiveness and getReadiness falling back to getStartups…

### LINKED ISSUE #15: HealthChecker.getLivenessStatus() and getReadinessStatus() returns aggregated status
We discovered an issue with https://github.com/strongloop/loopback-next/pull/3364. 

Please note `startupComplete` is only changed to `true` at https://github.com/CloudNativeJS/cloud-health/blob/master/src/healthcheck/HealthChecker.ts#L145 when `getStartupStatus()` is called, not upon the resolution of the startup check promise. As a result, calling `healthChecker.getLivenessStatus` and `healthChecker.getLivenessStatus` are falling back to [`getStartupStatus`](https://github.com/CloudNativeJS/cloud-health/blob/master/src/healthcheck/HealthChecker.ts#L182) if they are called first.

I have to work around the issue with https://github.com/strongloop/loopback-next/blob/health/packages/health/src/observers/health.observer.ts#L73.

### PR DESCRIPTION
**Pull request by andrewhughes101**
Fixes #15 
Added new tests to represent behaviour seen in #15 
Refactored getLiveness and getReadiness to no longer fall back to getStartupStatus 

cc @seabaylea  @raymondfeng

### CODE REVIEW

File: src/healthcheck/HealthChecker.ts [hunk 1]
```diff
@@ -118,51 +122,60 @@ class HealthChecker {
...
+    return this.getHealthStatus();
   }
 
+  private async getPromiseStatus(statusResponse: HealthStatus) {
+    this.startupPlugins.map((check) => {
+      const promiseCheck = check.runCheck();
+      statusResponse.addStatus(check.getStatus());
+      return promiseCheck;
+    });
+  }
```
Comment (raymondfeng) [inline src/healthcheck/HealthChecker.ts]: Should it be as follows?

```ts
const runChecks = this.startupPlugins.map(check => check.runCheck());
// Run all checks in parallel
await Promise.all(runChecks);
// Collect statuses
this.startupPlugins.forEach(check => statusResponse.addStatus(check.getStatus());
```

File: src/healthcheck/HealthChecker.ts [hunk 2]
```diff
@@ -118,51 +122,60 @@ class HealthChecker {
...
     if (this.startupComplete === false) {
-      statusResponse = new HealthStatus(State.UNKNOWN)
+      statusResponse = new HealthStatus(State.UNKNOWN);
 
-      this.startupPlugins.map((check) => {
-        const promiseCheck = check.runCheck();
-        statusResponse.addStatus(check.getStatus());
-        return promiseCheck;
-      });
+      this.getPromiseStatus(statusResponse);
```
Comment (raymondfeng) [inline src/healthcheck/HealthChecker.ts]: I think we should use:

```ts
await this.getPromiseStatus(statusResponse);
```

File: src/healthcheck/HealthChecker.ts [hunk 3]
```diff
@@ -118,51 +122,60 @@ class HealthChecker {
...
     let statusResponse: HealthStatus;
 
     if (this.shutdownRequested === true) {
-      return this.getShutdownStatus()
+      return this.getShutdownStatus();
     }
     if (this.startupComplete === false) {
-      return this.getStartupStatus()
+      statusResponse = new HealthStatus(State.UNKNOWN);
+      this.getPromiseStatus(statusResponse);
```
Comment (raymondfeng) [inline src/healthcheck/HealthChecker.ts]: same here - `await` is needed.

File: src/healthcheck/HealthChecker.ts [hunk 4]
```diff
@@ -174,20 +187,26 @@ class HealthChecker {
...
+    let statusResponse: HealthStatus;  
+
     if (this.shutdownRequested === true) {
-      return this.getShutdownStatus()
+      return this.getShutdownStatus();
     }
     if (this.startupComplete === false) {
-      return this.getStartupStatus()
+      statusResponse = new HealthStatus(State.UNKNOWN);
+      this.getPromiseStatus(statusResponse);
```
Comment (raymondfeng) [inline src/healthcheck/HealthChecker.ts]: same here - `await` is needed.

File: report.20190819.100201.56153.0.001.json [hunk 5]
```diff
@@ -0,0 +1,731 @@
+
```
Comment (richardlau) [inline report.20190819.100201.56153.0.001.json]: This file was accidentally added?

File: src/healthcheck/HealthChecker.ts [hunk 6]
```diff
@@ -118,51 +122,58 @@ class HealthChecker {
...
-      return this.getStartupStatus()
+      return this.getStartupStatus();
     }
-    return this.getHealthStatus()
+    return this.getHealthStatus();
+  }
+
+  private async getPromiseStatus(statusResponse: HealthStatus) {
+    const runChecks = this.startupPlugins.map(check => check.runCheck());
+    Promise.all(runChecks);
```
Comment (raymondfeng) [inline src/healthcheck/HealthChecker.ts]: `await` is missing here. Without `await`, the next statement will be executed by the `runCheck` is done. As a result, the statusResponse contains incorrect statuses.

File: src/healthcheck/HealthChecker.ts [hunk 7]
```diff
@@ -203,48 +220,48 @@ class HealthChecker {
...
 
     // Handle shutdown case
     if (this.shutdownRequested === true) {
-      statusResponse = new HealthStatus(State.STOPPING)
+      statusResponse = new HealthStatus(State.STOPPING);
       await this.shutdownPlugins.map((check) => {
         statusResponse.addStatus(check.getStatus());
-      })
-      return statusResponse
+      });
```
Comment (raymondfeng) [inline src/healthcheck/HealthChecker.ts:253]: It is suspicious. Please note the following code is executed synchronously and `check.runCheck()` is not invoked.

```ts
this.shutdownPlugins.map((check) => {
  statusResponse.addStatus(check.getStatus());
});
```

Comment (andrewhughes101) [reply src/healthcheck/HealthChecker.ts:253]: Added
``` JS
const runChecks = this.shutdownPlugins.map((check) => {
    check.runCheck();
 });
await Promise.all(runChecks);
this.shutdownPlugins.forEach((check) => {
    statusResponse.addStatus(check.getStatus());
});
```

File: src/healthcheck/HealthChecker.ts (hunk 6 above)
Comment (andrewhughes101) [reply src/healthcheck/HealthChecker.ts]: Originally left the `await` out as it failed quite a few tests. I have however, re-looked at the tests and made some changes. I have now added the `await`.

**Comment by raymondfeng** [comment]
@andrewhughes101 The latest version seems to be working fine now.

**Review by BethGriggs — APPROVED** [review]

**Comment by raymondfeng** [comment]
When can we expect a new release?

**Comment by BethGriggs** [comment]
@raymondfeng, @andrewhughes101 and I are looking at cutting a v2.1.0 release imminently

**Pull request merged**